File: src/parser_utils/export_condition_map.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Ast = Flow_ast

type 'a condition_value =
  | Null
  | Path of string
  | Nested of (string * 'a condition_value) list

type t = { conditions: (string * t condition_value) list }

let empty = { conditions = [] }

let create ~conditions = { conditions }

let create_from_shorthand_value ~value =
  let conditions = [("default", value)] in
  create ~conditions

let create_from_shorthand ~path = create_from_shorthand_value ~value:(Path path)

let is_targeted_condition valid_conditions candidate_condition =
  List.mem candidate_condition valid_conditions || candidate_condition = "default"

let rec pick_target valid_conditions pattern_match = function
  | (candidate_condition, Nested child_condition_map) :: rest ->
    if is_targeted_condition valid_conditions candidate_condition then
      match pick_target valid_conditions pattern_match child_condition_map with
      | Some t -> Some t
      | None -> pick_target valid_conditions pattern_match rest
    else
      pick_target valid_conditions pattern_match rest
  | (candidate_condition, Path target_path) :: rest ->
    if is_targeted_condition valid_conditions candidate_condition then
      match pattern_match with
      | Some pattern_match ->
        Some (Base.String.substr_replace_first ~pattern:"*" ~with_:pattern_match target_path)
      | None -> Some target_path
    else
      pick_target valid_conditions pattern_match rest
  | (_, Null) :: _ -> None
  | [] -> None

let resolve_package_target condition_map pattern_match valid_conditions =
  pick_target valid_conditions pattern_match condition_map.conditions

let rec parse_condition_property_value = function
  | Ast.Expression.NullLiteral _ -> Some Null
  | Ast.Expression.StringLiteral { Ast.StringLiteral.value; _ } -> Some (Path value)
  | Ast.Expression.Object { Ast.Expression.Object.properties; comments = _ } ->
    Some (Nested (parse_condition_map properties))
  | _ -> None

and parse_condition_property property =
  let open Ast in
  let open Expression.Object in
  match property with
  | Property
      ( _,
        Property.Init
          {
            key = Property.StringLiteral (_, { StringLiteral.value = key; _ });
            value = (_, value);
            _;
          }
      ) ->
    (match parse_condition_property_value value with
    | Some value -> Some (key, value)
    | None -> None)
  | _ -> None

and parse_condition_map properties = List.filter_map parse_condition_property properties

let parse = function
  | Ast.Expression.NullLiteral _ -> Some (create_from_shorthand_value ~value:Null)
  | Ast.Expression.StringLiteral { Ast.StringLiteral.value : string; _ } ->
    Some (create_from_shorthand ~path:value)
  | Ast.Expression.Object { Ast.Expression.Object.properties; comments = _ } ->
    let conditions = parse_condition_map properties in
    Some (create ~conditions)
  | _ -> None

File: rust_port/crates/flow_parser_utils/src/export_condition_map.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::ops::Deref;

use dupe::Dupe;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_parser::ast;
use flow_parser::ast::expression::ExpressionInner;
use flow_parser::loc::Loc;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ConditionValue {
    Null,
    Path(FlowSmolStr),
    Nested(Vec<(FlowSmolStr, ConditionValue)>),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ExportConditionMap {
    conditions: Vec<(FlowSmolStr, ConditionValue)>,
}

impl ExportConditionMap {
    pub fn empty() -> Self {
        ExportConditionMap {
            conditions: Vec::new(),
        }
    }

    pub fn create(conditions: Vec<(FlowSmolStr, ConditionValue)>) -> Self {
        ExportConditionMap { conditions }
    }

    fn create_from_shorthand_value(value: ConditionValue) -> Self {
        let conditions = vec![(FlowSmolStr::new("default"), value)];
        ExportConditionMap::create(conditions)
    }

    pub fn create_from_shorthand(path: FlowSmolStr) -> Self {
        ExportConditionMap::create_from_shorthand_value(ConditionValue::Path(path))
    }

    pub fn resolve_package_target(
        &self,
        pattern_match: Option<&str>,
        valid_conditions: &[FlowSmolStr],
    ) -> Option<FlowSmolStr> {
        pick_target(valid_conditions, pattern_match, &self.conditions)
    }

    pub fn parse(expr: &ast::expression::Expression<Loc, Loc>) -> Option<Self> {
        match expr.deref() {
            ExpressionInner::NullLiteral { .. } => Some(
                ExportConditionMap::create_from_shorthand_value(ConditionValue::Null),
            ),
            ExpressionInner::StringLiteral { inner, .. } => Some(
                ExportConditionMap::create_from_shorthand(inner.value.dupe()),
            ),
            ExpressionInner::Object { inner, .. } => {
                let conditions = parse_condition_map(&inner.properties);
                Some(ExportConditionMap::create(conditions))
            }
            _ => None,
        }
    }
}

fn is_targeted_condition(valid_conditions: &[FlowSmolStr], candidate_condition: &str) -> bool {
    valid_conditions
        .iter()
        .any(|c| c.as_str() == candidate_condition)
        || candidate_condition == "default"
}

fn pick_target(
    valid_conditions: &[FlowSmolStr],
    pattern_match: Option<&str>,
    conditions: &[(FlowSmolStr, ConditionValue)],
) -> Option<FlowSmolStr> {
    for (candidate_condition, value) in conditions {
        match value {
            ConditionValue::Null => {
                return None;
            }
            ConditionValue::Nested(child_condition_map) => {
                if is_targeted_condition(valid_conditions, candidate_condition.as_str()) {
                    if let Some(t) =
                        pick_target(valid_conditions, pattern_match, child_condition_map)
                    {
                        return Some(t);
                    }
                }
            }
            ConditionValue::Path(target_path) => {
                if is_targeted_condition(valid_conditions, candidate_condition.as_str()) {
                    return Some(match pattern_match {
                        Some(pattern_match) => {
                            FlowSmolStr::new(target_path.replacen('*', pattern_match, 1))
                        }
                        None => target_path.clone(),
                    });
                }
            }
        }
    }
    None
}

fn parse_condition_property_value(
    expr: &ast::expression::Expression<Loc, Loc>,
) -> Option<ConditionValue> {
    match expr.deref() {
        ExpressionInner::NullLiteral { .. } => Some(ConditionValue::Null),
        ExpressionInner::StringLiteral { inner, .. } => {
            Some(ConditionValue::Path(inner.value.dupe()))
        }
        ExpressionInner::Object { inner, .. } => Some(ConditionValue::Nested(parse_condition_map(
            &inner.properties,
        ))),
        _ => None,
    }
}

fn parse_condition_property(
    property: &ast::expression::object::Property<Loc, Loc>,
) -> Option<(FlowSmolStr, ConditionValue)> {
    if let ast::expression::object::Property::NormalProperty(
        ast::expression::object::NormalProperty::Init {
            key: ast::expression::object::Key::StringLiteral((_loc, lit)),
            value,
            ..
        },
    ) = property
    {
        if let Some(value) = parse_condition_property_value(value) {
            return Some((lit.value.dupe(), value));
        }
    }
    None
}

fn parse_condition_map(
    properties: &[ast::expression::object::Property<Loc, Loc>],
) -> Vec<(FlowSmolStr, ConditionValue)> {
    properties
        .iter()
        .filter_map(parse_condition_property)
        .collect()
}
